### PR TITLE
Revamp the contributor guide 

### PR DESCRIPTION
**Pull request by greg0ire**
While some sections feel redundant, other sections were missing, like
the one about choosing the right branch.

This only addresses the first question of #472 

@mpdude please review.

### CODE REVIEW

File: source/contribute.rst
Comment (SenseException) [inline source/contribute.rst:124]:
```suggestion
If you want to add a new feature, you should target the next minor
branch, because it might have unintentional side effects that break existing code.
```

**Review by SenseException — APPROVED** [review]
I'd like to see more reviewers for this.

**Comment by mpdude** [comment]
I‘ll have to review the branch selection rules later, with a clear mind.

In general, I'd say there is a lot of information in this chapter that should be routine for people familiar with `git`. 

Do we really need to provide all that? Do we anticipate contributions from users who haven’t done all that before?

**Comment by greg0ire** [comment]
> Do we really need to provide all that? Do we anticipate contributions from users who haven’t done all that before?

There's indeed pretty basic stuff, so we could remove more than I already did. That being said, here are the most common issues I see:

- people who develop directly on 2.3.x instead of creating a topic branch, and are then are asked to retarget on 2.4.x
- people who don't know they can force push and instead close their PR and open a new one when asked to retarget their branch
- bad commit messages (typically `fix for #1234`, where they don't explain what they did and how that's supposed to fix anything) 
- good commit message body, and commit message subject as suggested by github, so typically "Update path/to/file.php"
- non-atomic commits, with unexplained changed mixed with changes relating to the main topic of the pull request
- PR that passes the build, but with commits that don't pass the build individually
- commits that don't bring extra understanding of what's being done but rather show the struggle ("fix cs")

**Comment by greg0ire** [comment]
ping @mpdude :slightly_smiling_face:

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:66]: This should be familiar to anyone who has contributed on GitHub. For first time contributors, there is probably enough material out there on how to get started, including GitHub.com documentation. 

IMHO we do not need to write/maintain this information ourselves.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:73]: Does that apply to all Doctrine projects?

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:66]: Are you suggesting I remove this first paragraph, or replace it with a link? If the latter which one? This, maybe: https://docs.github.com/en/get-started/quickstart/github-flow ?

**Comment by mpdude** [comment]
Still have to go through the textual changes. 

Regarding the other points you brought up:

When we ask people to re-target PRs, can we expect/assume they find these instructions?

Would it help if we set up a canned reply in GitHub, so that when we ask people to change the target branch we could include a pointer to this documentation?

Regarding bad commit messages, meaningless commits etc.: Would it make sense to discuss squash-merging commits like Symfony does? That would help to hide “how the sausage is being made” and have a clean history with one commit per change (= PR).

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:66]: I'd say https://docs.github.com/en/get-started/quickstart/contributing-to-projects explains how to fork a repo, clone the fork, create a branch, push it and make a PR.

https://docs.github.com/en/get-started/quickstart/hello-world is even more basic and goes through GitHub fundamentals.

In general, it's really hard to tell what would help and invite people with little to no experience to contribute, when you've been using Git or GitHub for a long time. I tend to forget about how it was to learn.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:76]: What would be the best way to find out for a given Doctrine project what the current stable/bugfix/... version is?

Along the lines of...
--
You can find out at https://www.doctrine-project.org/projects/orm.html [<-- applies to ORM only] what the current stable as well as the upcoming minor and major versions are.

For example, let's say the current stable version is 2.3.1. This makes 2.4 as well as 3.0 the next minor and major version releases. In the repository, there will be corresponding branches:

Comment (greg0ire) [reply source/contribute.rst:76]: We sometimes forget to update the project metadata, so the best way is really to find what the latest (or rather, highest) version number is, and derive the branch information from that.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:94]: The general rules are as follows:
* Patch releases should be as stable as possible. They fix bugs, but in avoid making changes as far as possible.
* Improvements and additions are being made on the upcoming minor release branch. 
* An upgrade to a new minor version must not require any changes on the user's side.
* There must be a clear upgrade path from one major version to the next. That means necessary changes on the user's side will be advertised through deprecation notices, if anyway possible.
* Deprecations may be added for new minor versions only.
* Before a feature is removed (or an API changed) in a major version release, a corresponding deprecation notice must be included in a minor version release preceding it. 
* If a replacement feature and/or alternative API will be provided, it must be available in the minor release that adds the deprecation for the old feature.

And maybe one disputable extra item?

* When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

Comment (greg0ire) [reply source/contribute.rst]: > When a minor release brings (major) changes in behavior, ideally those would be opt-in to avoid surprises during the upgrade process.

That sounds like a breaking change, and wouldn't go in a minor release :thinking: 

I don't really understand what you mean I'm afraid.

Comment (mpdude) [reply source/contribute.rst]: I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project 🥶?

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:120]: The next minor version branch may include:
* Refactorings, unless they are necessary for bugfixes. This is to avoid unnecessary risks.
* Adding deprecations – remember to update the `UPGRADE.md` file as well!
* Adding new features and/or public APIs

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:126]: On the next major version branch, usually only deprecations notices, deprecated features and necessary compatibility layers are being removed.

Avoid suprising changes in public APIs on the next major version branch if anyhow possible. In order to provide a smooth upgrade path to users, the preferred solution is to add deprecations and possibly alternative APIs in a preceding minor version, and only remove the deprecations and compatibility layers in the next major release.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:131]: Same as before – do we really need this entire section?

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:120]: What about changing minimum dependency requirements?

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst]: That makes a lot more sense, yeah, let me add something about it.

Comment (greg0ire) [reply source/contribute.rst]: > I notice that this is not editing the contributor's documentation, but instead making policies for the Doctrine project cold_face?

Do you mean this feels out of place? Because https://www.doctrine-project.org/policies.html could be a good place for this.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:58]: There isn't AFAIK. I should create another PR to https://www.doctrine-project.org/contribute/maintainer/index.html I suppose.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:120]: Changing minimum dependency requirements is OK on a patch branch if it's absolutely necessary to fix a bug. I don't think that's often the case, and I think we don't have to mention this here.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:131]: For that one I know the answer, and that's yes. There are way too many contributors contributing directly to e.g. 2.14.x directly.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:329]: The paragraph is "typically because the build had an issue", and I personally don't approve pull requests that have a red build.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst]: I mean, who am I to write or direct rules that mandate how the Doctrine shall organize/maintain its code 😐

Comment (greg0ire) [reply source/contribute.rst]: Dunno… a future maintainer hopefully :crossed_fingers: ?

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:92]: Here we have two sentences that end with a colon

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:133]: Maybe it’s because we removed an initial section, but nothing in the above text supports the assumption that the remote might be called like this

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:139]: AFAIK these two are cache invalidation, naming things and off-by-one bugs

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:149]: Makes the same assumption

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:129]: Then, let's explain it...

You should avoid making changes directly on the patch, minor or major release branches, also in your fork of the repository. That makes it difficult for you to have more than one pull request at a time, and also complicates matters should you need to rebase your work to another branch some time later on. Instead, for every change that you would like to propose, create a dedicated topic branch. 

Note: If that does not sound all too familiar, you might want to read about GitHub Flow for additional background information https://docs.github.com/en/get-started/quickstart/github-flow.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:137]: Let's move that a bit down

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:227]: The base branch should be what you chose according to the guidelines mentioned in the beginning of this chapter. It is also the branch name that you provided when you created your topic branch.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:229]: That'a a bit hidden – on GitHub, you have to click "Edit" next to the pull request title to change the title and base branch

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:316]: Basically, a ebase takes all changes on your topic branch and moves them to another starting point. This starting point was the Doctrine branch that you chose when you created your topic branch, at that point in time. The rebase will move your changes to the current state of this branch.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:324]: Hm, that's pretty vague, and every line removed potentially is a lost commit. A nice footgun, eh?

As long as you're only rebasing on the same branch that you initially started out with, does it have to be an interactive rebase at all?

**Review by mpdude — COMMENTED** [review]
IMHO this is already in a very good shape, keep going!

Added a next round of feedback – sorry I missed some of the points last time.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:324]: (Assuming we're writing all this for people rather new to git and contributing on GitHub, so we should not talk them into complicated situations where they might lose work and/or get frustrated)

**Comment by mpdude** [comment]
Regarding https://github.com/doctrine/doctrine-website/pull/473#issuecomment-1403716535

> * bad commit messages (typically fix for #1234, where they don't explain what they did and how that's supposed to fix anything)
> * good commit message body, and commit message subject as suggested by github, so typically "Update path/to/file.php"
> * non-atomic commits, with unexplained changed mixed with changes relating to the main topic of the pull request
> * PR that passes the build, but with commits that don't pass the build individually
> * commits that don't bring extra understanding of what's being done but rather show the struggle ("fix cs")

IMHO squash-merging PRs could solve all those problems. We could also ask contributors to keep the initial PR description an up-to-date description and summary of the change and use that as the commit message for the merge. Symfony does it that way as well.

Where and how could this best be discussed with the relevant decision-makers from the Doctrine core team?

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:137]: Sorry, what exactly would you like me to move where, and for what reason?

Comment (greg0ire) [reply source/contribute.rst:137]: Disagree, see https://github.com/doctrine/doctrine-website/pull/473#discussion_r1130629469

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst]: Cool

Comment (mpdude) [reply source/contribute.rst]: Let’s have a dedicated section about changing the target branch and the rebase operations that are necessary for this. 

Additionally, prepare a canned reply that we can use in PRs whenever we ask people to change the target branch. This should point to that documentation, mention the correct branch and tell people that they _need not_ close the PR but can update it.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:169]: To be honest, the name chosen does not really matter much

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:179]: Ok so I think I addressed that.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst]: Addressed.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:49]: > Or to rename it… should be less confusing.

I think I misunderstood this, I thought you wanted to rename `origin` to something else.

I agree, there are people who clone the original repo and add their fork as a second remote, or they clone their fork and add `upstream` or similar.

Either way, wouldn't it be safe if we suggest to add a `doctrine` remote pointing to the original Doctrine project's repo, and use that remote name in all other git commands?

Comment (mpdude) [reply source/contribute.rst:49]: I just see that in fact, you do that – rename `origin` to `doctrine`. That may be wrong, depending on how people set up their clone. And if they follow the GitHub instruction linked above, it will definetly be wrong, since those instructions end up with `origin`= the fork.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:179]: Great

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:197]: Starting from "In fact..." – do we really want to suggest this, and is it relevant in the context of creating a feature branch and committing on this?

Yes, it's an additional safeguard, but nothing you need to repeat every time you create a new branch.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:254]: `fork`? 🤔

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:279]: Relevant information?

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:368]: 0. Switch to the branch you'd like to rebase.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:385]: Good point, I was not aware of that!

Comment (mpdude) [reply source/contribute.rst:385]: Wait, subconsciously I was.

```
git checkout my-topic
git rebase doctrine/2.3.x --onto doctrine/2.4.x
```

No need to pick the right commits. Am I missing something?

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:353]: "on" or "onto"? Not a native speaker myself. Git CLI suggests `onto`.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:403]: 6. Update the Pull Request target in GitHub

**Comment by mpdude** [comment]
Sorry for all the nitpicking, I owe you a beer once we happen to meet IRL

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:197]: Let's remove it, it makes the guide longer.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:279]: Are you asking if this sentence is relevant, or do you want me to add relevant information?

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:371]: Yes.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:385]: You're probably right, let's just suggest that one method.

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:353]: From `man git rebase`: "it may not be the best idea to keep rebasing on top of the upstream"

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:403]: Already mentioned above (and I think it's best to do it before pushing)

**Comment by greg0ire** [comment]
Can't wait for the next FOSSA-EU hackathon!

**Comment by mpdude** [comment]
Oh good ol' times

**Comment by greg0ire** [comment]
@mpdude IMO this is ready, can I get a formal approval from you? @SenseException , would it be sufficient for you or do you wish to get other maintainers to review this? I'd hate to late this rot…

**Comment by SenseException** [comment]
Yes, @mpdude fits perfectly as second reviewer. His approval should make this getting merged. :+1:

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:49]: @greg0ire I think this point still needs to be addressed:

Simply renaming `origin` to `doctrine` may not be correct. Depending on how people obtained their clone, `origin` may be their fork or the original (upstream) repo.

When closely following the instructions linked in the previous sentence, then `origin` will be the _fork_, so the rename suggested here is definetly wrong.

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:65]: Suggestion for this entire section:

---

To make the example commands given in the rest of this page less ambiguous, we will add two explicit names for remote Git repositories: `doctrine` for the original Doctrine project you're contributing to, and `fork` for your fork of it.

.. code-block:: console
    $ [`cd` into the directory created by `git clone ...`]
    $ git remote add doctrine git@github.com:doctrine/<name-of-project>.git
    $ git remote add fork git@github.com:<your-github-name>/<name-of-project>.git

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:173]: We don't need this note since we've added the remote name in the beginning section.

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:247]: This is now correct :+1:

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:362]: Is this missing a backtick?

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:380]: When using that method, the "commits you did not author" thingy is not a issue at all, right? If so, we need not mention it.

---

Another case where you need a rebase is when you want to change the
target branch of your PR. For instance, you might have created your PR
against ``2.3.x`` but you are told to change it to ``2.4.x``. In that
case, the following command will pick all changes that you made against the ``2.3.x`` branch, and re-apply them on the current ``2.4.x`` branch.

$ git fetch doctrine
$ git rebase --onto doctrine/2.4.x doctrine/2.3.x your-topic-branch
$ git push --force

**Comment by mpdude** [comment]
@greg0ire Done. Can you easily follow up on the new comments I made? Otherwise, I'll walk you through them one by one.

Not much left – let's push this over the finish line soon!

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:380]: > When using that method, the "commits you did not author" thingy is not a issue at all, right? If so, we need not mention it.

Not an issue indeed 👍

File: source/contribute.rst
Comment (greg0ire) [reply source/contribute.rst:362]: It is 👍

**Comment by greg0ire** [comment]
>  Can you easily follow up on the new comments I made? Otherwise, I'll walk you through them one by one.

I think I did, but I also think I might have easily missed some comments. Can you mark the resolved ones as such?

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:76]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:79]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:149]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:181]: ✅

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:191]: ✅

**Comment by mpdude** [comment]
For some reason, I cannot mark (everything) as "resolved" 🤔 .

Let's see:

- [ ] https://github.com/doctrine/doctrine-website/pull/473#discussion_r1166559279

File: source/contribute.rst
Comment (mpdude) [inline source/contribute.rst:62]: .rst need double backticks here and in the following lines?

File: source/contribute.rst
Comment (mpdude) [reply source/contribute.rst:277]: ✅

**Review by mpdude — APPROVED** [review]

**Comment by mpdude** [comment]
All comments addressed ✅ .

Awesome work @greg0ire 🥇 and sorry it took so many iterations. I really like the result!

**Comment by greg0ire** [comment]
Thanks for the thorough review!

**Pull request merged**
